### PR TITLE
add inter-process coordination

### LINKED ISSUE #10: Feature Request: Signal `tockloader listen` processes
In my experience any tockloader command will fail if another terminal has `tockloader listen` running. Therefore it would be nice if on any command, tockloader would signal all other tockloader processes, and if those were `tockloader listen` on the same port, they would release the serial port, wait for the signalling process to finish, and then resume listen. This would make tockloader work seamlessly with itself.

Note: I'm not actually sure what the best way to get tockloader processes to communicate is.

### PR DESCRIPTION
**Pull request by ppannuto**
The approach we take is to have every tockloader instance either
connect to a server named uniquely for each serial port, or if one
doesn't exist make that server

For most operations, the server will just reject other tockloader
instances. If the running process is a `listen`, however, we want
to essentially pause the listen while the other operation goes through
and then resume again. Unfortunately, due the the way that miniterm
is architected, this proves much easier to do by killing off the
current listen process and spinning up a new one, with a bonus flag
that will have it wait for the other running tockloader operation to
complete

All of these server coordination things take place during the
construction of the serial port access, so it shouldn't affect
commands that don't touch serial.

Tested reasonably extensively with hail and OS X.
Closes #10.

### CODE REVIEW

File: tockloader/main.py [hunk 1]
```diff
@@ -899,11 +882,137 @@ def open_link_to_board (self):
...
+			args = list(sys.argv)
+			args.append('--wait-to-listen')
+			os.execvp(args[0], args)
+
+
 	def get_serial_port (self):
 		return self.sp
 
+	# Get an identifier that will be consistent for this serial port on this
+	# machine that is also gaurenteed to not have any special characters (like
```
Comment (bradjc) [inline tockloader/main.py]: gaurenteed -> guaranteed

File: tockloader/main.py [hunk 2]
```diff
@@ -899,11 +882,137 @@ def open_link_to_board (self):
...
+			os.unlink(self.comm_path)
+			# Need to run miniterm's atexit handler
+			self.miniterm.console.cleanup()
+
+			args = list(sys.argv)
+			args.append('--wait-to-listen')
+			os.execvp(args[0], args)
+
+
 	def get_serial_port (self):
```
Comment (bradjc) [inline tockloader/main.py]: All instances of `get_serial_port` can be deleted now that `run_terminal` moved to this class.

File: tockloader/main.py [hunk 3]
```diff
@@ -1239,6 +1348,33 @@ def determine_current_board (self):
 			raise TockLoaderException('Could not determine the current board or arch')
 
 
+	# Run miniterm for receiving data from the board.
+	def run_terminal(self):
+		print('Listening for serial output.')
+
+		# Use trusty miniterm
+		self.miniterm = serial.tools.miniterm.Miniterm(
+			self.get_serial_port(),
```
Comment (bradjc) [inline tockloader/main.py]: This can just be `self.sp`

File: tockloader/main.py [hunk 4]
```diff
@@ -899,11 +882,137 @@ def open_link_to_board (self):
...
+		# Only one process at a time can talk to a serial port (reliably)
+		# Before connecting, check whether there is another tockloader process
+		# running, if it's a listen, pause listening, otherwise bail out
+		self.comm_path = '/tmp/tockloader.' + self._get_serial_port_hash()
+		if os.path.exists(self.comm_path):
+			self.client_sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+			try:
+				self.client_sock.connect(self.comm_path)
+			except ConnectionRefusedError:
+				print("  [Warning]: Found stale tockloader server, removing")
```
Comment (bradjc) [inline tockloader/main.py]: `"` -> `'`

**Comment by ppannuto** [comment]
Should all be fixed

**Pull request merged**